x/programs/rust/examples/token/src/lib.rs

```diff
@@ -25,27 +38,38 @@ pub fn mint_to(state: State, recipient: Address, amount: i64) -> bool {
         .is_ok()
 }
 
-/// Transfers amount coins from the sender to the recipient. Returns whether successful.
+/// @notice Transfers coins from the sender to the recipient
+/// @param state The current program state
+/// @param sender The address sending tokens
+/// @param recipient The address receiving tokens
+/// @param amount The amount of tokens to transfer
+/// @return Whether the transfer succeeded
 #[public]
 pub fn transfer(state: State, sender: Address, recipient: Address, amount: i64) -> bool {
     // require sender != recipient
-    if sender == recipient {
-        return false;
-    }
+    assert_ne!(sender, recipient);
+
     // ensure the sender has adequate balance
     let sender_balance: i64 = state.get_map_value("balances", &sender).unwrap_or(0);
-    if amount < 0 || sender_balance < amount {
-        return false;
-    }
+    assert!(amount >= 0 && sender_balance >= amount);
```

Comment:
before this change invalid balance returned false now it panics? just making sure I understand the intent here.

Reply:
can you compile and test this case I am curious how the engine handles this.

Reply:
This is closer to how the EVM handles these sorts of things fwiw (reverting out as soon as an assert is hit with no return).

Reply:
OK gothca, looking around I see Soroban does something similar WRT to `panic!`. But also adds context to the panic condition. Perhaps this could be a middle ground? Ideally we have something to return to the caller?

https://github.com/stellar/soroban-examples/blob/7a7cc6268ada55113ce0b82a3ae4405f7ec8b8f0/token/src/balance.rs#L29-L34

Reply:
Also important is that they explicitly define panic = abort to release which appears to have some interesting properties.

```
[profile.release]
panic = 'abort'
```

[1] https://github.com/stellar/soroban-examples/blob/7a7cc6268ada55113ce0b82a3ae4405f7ec8b8f0/token/Cargo.toml#L23
[2] https://doc.rust-lang.org/book/ch09-01-unrecoverable-errors-with-panic.html#unwinding-the-stack-or-aborting-in-response-to-a-panic

Reply:
Should probably add a custom `panic` handler:
https://doc.rust-lang.org/nomicon/panic-handler.html